docs/source/optimisation.rst

```diff
@@ -424,9 +404,54 @@ In addition, we provide a random sampling class which is a child class of  `cil.
 .. autoclass:: cil.optimisation.utilities.SamplerRandom
    :members:
 
+Callbacks
+---------
+
+A list of :code:`Callback` s to be executed each iteration can be passed to `Algorithms`_ :code:`run` method.
+
+.. code-block :: python
+
+   from cil.optimisation.utilities.callbacks import LogfileCallback
+   ...
+   algorithm.run(..., callbacks=[LogfileCallback("log.txt")])
+
+.. autoclass:: cil.optimisation.utilities.callbacks.Callback
+   :members:
+
+.. autoclass:: cil.optimisation.utilities.callbacks.ProgressCallback
+   :members:
+
+.. autoclass:: cil.optimisation.utilities.callbacks.TextProgressCallback
+   :members:
+
+.. autoclass:: cil.optimisation.utilities.callbacks.LogfileCallback
+   :members:
+
+Example use for early stopping:
+
+.. code:: python
+
+   from cil.optimisation.algorithms import Algorithm
+   from cil.optimisation.utilities import callbacks
+
+   class TestAlgo(Algorithm):
+        def __init__(self, *args, **kwargs):
+            super().__init__(*args, ***kwargs)
+            self.configured = True
+
+        def update(self):
+            self.x = -self.iteration
 
+        def update_objective(self):
+            self.loss.append(2 ** getattr(self, 'x', np.nan))
 
+      class EarlyStopping(callbacks.Callback):
+         def __call__(self, algorithm: Algorithm):
+               if algorithm.x <= -15:  # arbitrary stopping criterion
+                  raise StopIteration
 
+      algo = self.TestAlgo()
+      algo.run(20, callbacks=[callbacks.ProgressCallback(), EarlyStopping()])
```

Comment:
Can we show the output here? Would be good to see it terminating after 3 iterations